torch/nn/utils/attention.py

```diff
@@ -0,0 +1,274 @@
+""" Define Base class as well as some crowd favorites """
+from abc import ABC, abstractmethod
+from enum import IntEnum
+from typing import Optional, Union
+from warnings import warn
+
+import torch
+from torch.backends.cuda import (
+    can_use_efficient_attention,
+    can_use_flash_attention,
+    SDPAParams,
+)
+from torch.nn.functional import scaled_dot_product_attention
+from torch.utils import _pytree as pytree
+
+
+def input_requires_grad(*tensors: torch.Tensor) -> bool:
+    return any(t.requires_grad for t in tensors)
+
+
+def materialize_if_needed(
+    bias: "AttnBias", device: Optional[torch.device] = None
+) -> Union[torch.Tensor, "AttnBias"]:
+    if bias.needs_materialization():
+        return bias.materialize(device)
+    return bias
+
+
+def postprocess_flash_output(inpt_tensor: torch.Tensor, og_size: int):
+    """Handles the unpad of the last dimension"""
+    if inpt_tensor.size(-1) != og_size:
+        return inpt_tensor[..., :og_size]
+    return inpt_tensor
+
+
+class AttnBias(ABC):
+    """Abstract base class for attention biases"""
+
+    @abstractmethod
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+    @abstractmethod
+    def needs_materialization(self) -> bool:
+        raise NotImplementedError("This is an abstract base class")
+
+    @staticmethod
+    @abstractmethod
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask,
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+
+class TensorBias(AttnBias):
```

Comment:
originally I was going to have this because I wanted to push all people to using this for creating attention biases. Now that we are using __torch__function__ I don't know if the added overhead makes sense. As well I have concerns around compile + export + every other subsystem

cc @albanD

Reply:
I think your statement hit captured it perfectly, is the problem that this isn't explicitly stated ?

Reply:
In terms of overhead I meant the overhead of torch__function, which I haven't measured soo little preamture. But there is some pytree maps and I imagine for some they wouldn't wanna pay this overhead if they are using attention_masks that is just a tensor